riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1822 @@
+    {
+        // wrap_bit is used as sign_bit here.
+        tmp1_l = X_b1 + tmp2_l * 2**8 + wrap_bit * 2**15,
+        tmp3_h = wrap_bit * 0xffff,
+        tmp3_l = tmp1_l
+    }
+
+    // ======================= assertions =========================
+
+    instr fail 
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, 0, STEP)
+    {
+      tmp1_h = 1
+    }
+
+    // Computes Q = val(Y) / val(X) and R = val(Y) % val(X) and stores them in registers Z and W.
+    instr divremu YL, XL, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, YL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, XL, STEP + 1)
+        link if (1 - XXIsZero) ~> (tmp4_h, tmp4_l, tmp3_h, tmp3_l) = arith_mul.div(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link if (1 - XXIsZero) ~> wrap_bit = add_sub.gt(tmp3_h, tmp3_l, tmp2_h, tmp2_l)
+        link if (1 - XXIsZero) ~> wrap_bit_2 = add_sub.gt(tmp3_h, tmp3_l, 0, 0)
+        link ~> regs.mstore(0, ZL, STEP + 2, tmp6_h, tmp6_l)
+        link ~> regs.mstore(0, WL, STEP + 3, tmp5_h, tmp5_l)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = tmp2_h + tmp2_l,
+
+        // if tmp2 (the divisor) is zero, remainder is set to dividend, as per RISC-V specification:
+        XXIsZero * tmp1_h + (1 - XXIsZero) * tmp3_h = tmp5_h,
+        XXIsZero * tmp1_l + (1 - XXIsZero) * tmp3_l = tmp5_l,
+
+        // remainder >= 0:
+        (1 - XXIsZero) * wrap_bit_2 = 0, 
+
+        // remainder < divisor, conditioned to val(X) not being 0:
+        (1 - XXIsZero) * (wrap_bit - 1) = 0,
+
+        // in case X is zero, we set quotient according to RISC-V specification
+        XXIsZero * 0xffff + (1 - XXIsZero) * tmp4_h = tmp6_h,
+        XXIsZero * 0xffff + (1 - XXIsZero) * tmp4_l = tmp6_l
+    }
+"# + mul_instruction
+}
+
+fn mul_instruction() -> &'static str {
+    r#"
+    // Computes V = val(X) * val(Y) and
+    // stores the lower 32 bits in register Z and the upper 32 bits in register W.
+    instr mul XL, YL, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> (tmp4_h, tmp4_l, tmp3_h, tmp3_l) = arith_mul.mul(tmp1_h, tmp1_l, 0, 0, tmp2_h, tmp2_l)
+        link ~> regs.mstore(0, ZL, STEP + 2, tmp3_h, tmp3_l)
+        link ~> regs.mstore(0, WL, STEP + 3, tmp4_h, tmp4_l);
+"#
+}
+
+fn memory(with_bootloader: bool) -> String {
+    let memory_machine = if with_bootloader {
+        todo!()
+    } else {
+        r#"
+    std::machines::small_field::memory::Memory memory(bit12, byte2);
+"#
+    };
+
+    memory_machine.to_string()
+        + r#"
+
+    // Increased by 4 in each step, because we do up to 4 register memory accesses per step
+    col fixed STEP(i) { 4 * i };
+
+    // ============== memory instructions ==============
+
+    /// Loads one word from an address V = val(X) + Y and rounds it down to the next multiple of 4.
+    /// Writes the loaded word and the remainder of the division by 4 to registers Z and W,
```

Comment:
```suggestion
    /// Writes the loaded word and the remainder of the division by 4 to registers ZL and WL,
```

I don't know if these are left here on purpose